### ISSUE TITLE
Flexible price of NodeProvider

### ISSUE THREAD

**Issue by tuler** [sunodo/sunodo#329]
## 📚 Context

The price policy implemented by the [NodeProvider](https://github.com/sunodo/sunodo/blob/main/packages/contracts/contracts/provider/NodeProvider.sol) and the [Vault](https://github.com/sunodo/sunodo/blob/main/packages/contracts/contracts/payment/Vault.sol) implementation is a fixed price, meaning once the price is defined at instantiation time it cannot be changed.

This requires that the service provider sets the "correct" price right from the start, a value that is attractive to users and economically sustainable at the same time. This can be very hard, especially with the architecture of the node still changing.

The idea behind not allowing the service provider to raise the price was to protect users from price strategies that may be unfair to users. Like starting with a zero price to attract customers, provoking a lock-in, and then hiking up the price.

We should always avoid lock-in, by allowing users to change providers. This is already supported by changing the Authority consensus (which the application owner has the power to do) and register an existing application to a different provider. We still don't support withdrawing funds that were already committed to a service provider, but that could implemented in the future.

Regarding the price policy concerns, the trust model between the application developer and the validation service provider, especially in the Authority phase we are in, already requires a high degree of trust, for application state claims, and a pre-paid execution that does not provide guarantees that the service will be executed properly, with no slashing mechanism.

## ✔️ Solution

Given the current trust model I suggest we let the service provide freely change the price of the service.
This will give more freedom in the beginning phase of the service.

## 📈 Subtasks

- [ ] change the [Vault](https://github.com/sunodo/sunodo/blob/main/packages/contracts/contracts/payment/Vault.sol) smart contract to allow price changes. Emit an event so we can index it later.

**Comment by tuler** [comment]
RFC @guidanoli @felipeargento @pedroargento

**Comment by guidanoli** [comment]
A measure to protect users would be to delay price increases by, say, 1 week, and allow users to withdraw funds committed to validators.

**Comment by tuler** [comment]
I think the additional complexity of that model is not worth it at this phase. I would prefer a simpler model.

**Comment by pedroargento** [comment]
If we don't add complexity to the model there is very little difference between the current implementation (with the possibility of price increase) and just moving the runway calculation to the controller and the vault being just a vault. Its the same risks and exposure for who's financing the DApp.

Maybe one solution could be changing the logic of the runway calculation. Instead of keeping track of the number of seconds available, keep track of the expiration date. When money is deposited, it buys certain amount of seconds based on the current price and extends the expiration. The validator can choose as a parameter the maximum expiration timestamp it want to commit at the current price. This way, a new price does not affect already bought seconds.

Something like that
```
function _extendRunway(
        address _msgSender,
        ICartesiDApp _application,
        uint256 _value
    ) internal returns (uint256) {
        uint256 secondsBought = _value/price;
        require(secondsBought <= maximumRunwayExtension)
        token.safeTransferFrom(_msgSender, payee, value);
        Expiration = expiration + secondsBought
        
        emit FinancialRunway(address(_application), Expiration);

        return secondsBought;
    }

```

**Comment by tuler** [comment]
> Maybe one solution could be changing the logic of the runway calculation. Instead of keeping track of the number of seconds available, keep track of the expiration date. When money is deposited, it buys certain amount of seconds based on the current price and extends the expiration. The validator can choose as a parameter the maximum expiration timestamp it want to commit at the current price. This way, a new price does not affect already bought seconds.

This is exactly what I already do.
Hiking up the price does not affect what was already deposited.

**Comment by pedroargento** [comment]
That is true, I misinterpreted the meaning of runway as a delta time instead of expiration. Then, I don't think its a big deal changing the price, the only thing the validator can do is change the price to `maxInt` and take all the deposited money at once, but the runway will still show the correct expiration, making it easier to prove the node provider is not fulfilling his end of the deal.

But I think its worth to think about of giving the option to limit renewals otherwise it can be moot to increase the price if the DApp already bought years of runway. This way a validator could express how much risk he wants to incur.

**Comment by pedroargento** [comment]
I would enforce it onchain. I know there is already a lot of offchain trust in the process, but putting this kind of thing onchain can allow people in the future to monitor the blockchain and create a reputation score for validators in the market place.

If this score does exist, a DApp not following the polite ask to not buy too much time at once will hurt the validator that unilaterally stops validating before runway ends because  the price is outdated.

**Comment by tuler** [comment]
> a DApp not following the polite ask to not buy too much time at once will hurt the validator that unilaterally stops validating before runway ends because the price is outdated.

But the validator knows that whatever was already deposited must be honored, and price changes only affects new deposits.
So if the validator unilaterally stops validating it is in fact breaking the "contract", and must hurt his reputation.

**Comment by pedroargento** [comment]
Yes. Thats why I think its a good feature to allow the validator to limit deposits to only extend the runway at the current price up to a point he is comfortable with the risk.

**Comment by tuler** [comment]
> Thats why I think its a good feature to allow the validator to limit deposits

What is the default, or suggested limit in your opinion?

**Comment by pedroargento** [comment]
I would say the default should be no limit. The suggested should be  based on validator cost volatility and trend. 

I would give the opportunity to set this limit together with the price increase function. Then, as the node becomes more mature and the costs more well known, the validator can adjust the limit.

**Comment by felipeargento** [comment]
I like the solution described.
It strikes a good balance between simplicity and transparency/accountability. I think emitting an event is a must.

I agree with the idea of limiting the amount of runway the dapp can buy on a specific price point too. Otherwise the risk for validators can be huge. I'd suggest something like 6 months or 1 year...so you only let deposits that extend the runaway up to a year from it's date. Can even just transfer the rest of the money back, if you don't want to reverse the transaction.

Another thing that comes to mind is: is there a way to charge different prices from different dapps? We talked about doing that for different inputs, but thats the dapp charging.

I think in the future a tool that simulates da dapp with some random inputs and figures out the cost of running it would be interesting. Especially when solutions like what @milton suggested on Discord become a reality (composability, mega dapps etc). But maybe thats for a future discussion.

**Comment by tuler** [comment]
In terms of different prices my idea is to have multiple providers, with different prices, that the user can choose.
The criteria that comes to mind is required resources, like the maximum size of the Cartesi machine.
Sunodo will need to impose some limit on the size for sure.
